tree/ntuple/v7/src/RNTupleSerialize.cxx

```diff
@@ -739,23 +801,29 @@ std::uint32_t ROOT::Experimental::Internal::RNTupleSerializer::SerializeLocator(
    const RNTupleLocator &locator, void *buffer)
 {
    std::uint32_t size = 0;
-   if (!locator.fUrl.empty()) {
-      if (locator.fUrl.length() >= (1 << 24))
+   if (locator.IsSimple()) {
+      if (static_cast<std::int32_t>(locator.fBytesOnStorage) < 0)
          throw RException(R__FAIL("locator too large"));
-      std::int32_t head = locator.fUrl.length();
-      head |= 0x02 << 24;
-      head = -head;
-      size += SerializeInt32(head, buffer);
-      if (buffer)
-         memcpy(reinterpret_cast<unsigned char *>(buffer) + size, locator.fUrl.data(), locator.fUrl.length());
-      size += locator.fUrl.length();
+      size += SerializeUInt32(locator.fBytesOnStorage, buffer);
+      size += SerializeUInt64(locator.Get<std::uint64_t>(),
+                              buffer ? reinterpret_cast<unsigned char *>(buffer) + size : nullptr);
       return size;
    }
 
-   if (static_cast<std::int32_t>(locator.fBytesOnStorage) < 0)
-      throw RException(R__FAIL("locator too large"));
-   size += SerializeUInt32(locator.fBytesOnStorage, buffer);
-   size += SerializeUInt64(locator.fPosition, buffer ? reinterpret_cast<unsigned char *>(buffer) + size : nullptr);
+   const int type = std::holds_alternative<std::string>(locator.fPosition)
+                       ? static_cast<std::uint8_t>(ROOT::Experimental::RNTupleLocator::kTypeURI)
+                       : locator.fType;
+   if (!type)
+      throw RException(R__FAIL("locator has no type"));
+   auto payloadp = buffer ? reinterpret_cast<unsigned char *>(buffer) + sizeof(std::int32_t) : nullptr;
+   size += std::visit(
+      [&](auto &&value) { return SerializeLocatorPayload<std::decay_t<decltype(value)>>(locator, payloadp); },
+      locator.fPosition);
+   std::int32_t head = sizeof(std::int32_t) + size;
+   head |= locator.fReserved << 16;
+   head |= type << 24;
+   head = -head;
```

Comment:
I wonder: is this consistent with the specification? We could also cut off the sign bit. I actually like the `*(-1)` trick because I find the (de-)serialization code simpler to read that way. We should just make sure then that the specification is clear about it.